From a0c70f4184623bd5f12a8a8a9e62e75f71c6bfa0 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Wed, 29 Jun 2011 17:57:26 -0600
Subject: [PATCH] Add a "Save" button to the PatchScriptSettingsPanel

The "Update" button now only updates the display. For
non-logged in users it behaves as it used to.  However,
for logged in users, the "Update" button no longer saves
the user preferences, and all of its settings are reset
to the saved values whenenver the PatchScreen is reloaded
after the update.  Additionaly, for logged in users, a
"Save" button now appears to the right of the "Update"
button and now behaves the way that "Update" used to
behave for logged in users.

Change-Id: I6ee5d688fbcee268aef1de1408762904d2a3c1ea
---
 .../google/gerrit/client/patches/PatchScreen.java  |    3 ++
 .../client/patches/PatchScriptSettingsPanel.java   |   22 ++++++++++++----
 .../client/patches/PatchScriptSettingsPanel.ui.xml |   12 ++++++++-
 .../client/ui/ListenableAccountDiffPreference.java |   25 ++++++++++++++++++++
 4 files changed, 54 insertions(+), 8 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
index 70df33e..910bf61 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
@@ -153,6 +153,9 @@ public abstract class PatchScreen extends Screen implements
 
     prefs = fileList != null ? fileList.getPreferences() :
                                new ListenableAccountDiffPreference();
+    if (Gerrit.isSignedIn()) {
+      prefs.reset();
+    }
     prefs.addValueChangeHandler(
         new ValueChangeHandler<AccountDiffPreference>() {
           @Override
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
index 395ccdb..70aade1 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
@@ -96,6 +96,9 @@ public class PatchScriptSettingsPanel extends Composite implements
   @UiField
   Button update;
 
+  @UiField
+  Button save;
+
   /**
    * Counts +1 for every setEnabled(true) and -1 for every setEnabled(false)
    *
@@ -117,13 +120,14 @@ public class PatchScriptSettingsPanel extends Composite implements
     initContext(context);
     if (!Gerrit.isSignedIn()) {
       reviewed.setVisible(false);
+      save.setVisible(false);
     }
 
     KeyPressHandler onEnter = new KeyPressHandler() {
       @Override
       public void onKeyPress(KeyPressEvent event) {
         if (event.getNativeEvent().getKeyCode() == KeyCodes.KEY_ENTER) {
-          update();
+          save();
         }
       }
     };
@@ -162,7 +166,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     } else {
       syntaxHighlighting.setValue(false);
     }
-    toggleEnabledStatus(update.isEnabled());
+    toggleEnabledStatus(save.isEnabled());
   }
 
   public void setEnableIntralineDifference(final boolean on) {
@@ -172,7 +176,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     } else {
       intralineDifference.setValue(false);
     }
-    toggleEnabledStatus(update.isEnabled());
+    toggleEnabledStatus(save.isEnabled());
   }
 
   private void toggleEnabledStatus(final boolean on) {
@@ -223,6 +227,11 @@ public class PatchScriptSettingsPanel extends Composite implements
     update();
   }
 
+  @UiHandler("save")
+  void onSave(ClickEvent event) {
+    save();
+  }
+
   private void update() {
     if (colWidth.getIntValue() <= 0) {
       new ErrorDialog(PatchUtil.C.illegalNumberOfColumns()).center();
@@ -244,7 +253,10 @@ public class PatchScriptSettingsPanel extends Composite implements
     dp.setRetainHeader(retainHeader.getValue());
 
     listenablePrefs.set(dp);
+  }
 
+  private void save() {
+    update();
     if (Gerrit.isSignedIn()) {
       persistDiffPreferences();
     }
@@ -252,11 +264,9 @@ public class PatchScriptSettingsPanel extends Composite implements
 
   private void persistDiffPreferences() {
     setEnabled(false);
-    Util.ACCOUNT_SVC.changeDiffPreferences(getValue(),
-        new GerritCallback<VoidResult>() {
+    listenablePrefs.save(new GerritCallback<VoidResult>() {
       @Override
       public void onSuccess(VoidResult result) {
-        Gerrit.setAccountDiffPreference(getValue());
         setEnabled(true);
       }
 
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml
index 3ec5467..f320ce5 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml
@@ -147,7 +147,8 @@ limitations under the License.
       </g:CheckBox>
     </td>
 
-    <td valign='bottom' rowspan='2'>
+    <td rowspan='2'>
+      <br/>
       <g:Button
           ui:field='update'
           text='Update'
@@ -155,13 +156,20 @@ limitations under the License.
           tabIndex='13'>
         <ui:attribute name='text'/>
       </g:Button>
+      <g:Button
+          ui:field='save'
+          text='Save'
+          styleName='{style.updateButton}'
+          tabIndex='14'>
+        <ui:attribute name='text'/>
+      </g:Button>
     </td>
 
     <td>
       <g:CheckBox
           ui:field='reviewed'
           text='Reviewed'
-          tabIndex='14'>
+          tabIndex='15'>
         <ui:attribute name='text'/>
       </g:CheckBox>
     </td>
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
index 195adf5..41de2cd 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
@@ -15,12 +15,37 @@
 package com.google.gerrit.client.ui;
 
 import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.account.Util;
+import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.reviewdb.AccountDiffPreference;
+import com.google.gwtjsonrpc.client.VoidResult;
 
 public class ListenableAccountDiffPreference
     extends ListenableValue<AccountDiffPreference> {
 
   public ListenableAccountDiffPreference() {
+    reset();
+  }
+
+  public void save(final GerritCallback<VoidResult> cb) {
+    if (Gerrit.isSignedIn()) {
+      Util.ACCOUNT_SVC.changeDiffPreferences(get(),
+          new GerritCallback<VoidResult>() {
+        @Override
+        public void onSuccess(VoidResult result) {
+          Gerrit.setAccountDiffPreference(get());
+          cb.onSuccess(result);
+        }
+
+        @Override
+        public void onFailure(Throwable caught) {
+          cb.onFailure(caught);
+        }
+      });
+    }
+  }
+
+  public void reset() {
     if (Gerrit.isSignedIn() && Gerrit.getAccountDiffPreference() != null) {
       set(Gerrit.getAccountDiffPreference());
     } else {
-- 
1.7.3.1

